tests/GeneratedHydratorTest/CodeGenerator/Visitor/HydratorMethodsVisitorTest.php

```diff
@@ -81,4 +83,106 @@ public function classAstProvider() : array
             [$staticClassName, $parser->parse('<?php ' . $staticClassCode)[0], ['taz']],
         ];
     }
+
+	/**
+	 * @dataProvider propertyAstProvider
+	 */
+	public function testPropertyCodeGeneration(string $className, Class_ $classNode, array $properties) : void
+	{
+		$visitor = new HydratorMethodsVisitor(new ReflectionClass($className));
+
+		/** @var Class_ $modifiedAst */
+		$modifiedNode = $visitor->leaveNode($classNode);
+
+		$constructors = $this->findConstructor($modifiedNode);
+		self::assertCount(1, $constructors);
+		$constructor = reset($constructors);
+
+		$hydratePropertyNames = $this->findAssignedPropertyNames(
+			$constructor,
+			'hydrateCallbacks',
+			'object',
+			function (Assign $assign) {
+				return $assign->var->name->name;
+			}
+		);
+		$extractPropertyNames = $this->findAssignedPropertyNames(
+			$constructor,
+			'extractCallbacks',
+			'values',
+			function (Assign $assign) {
+				return $assign->var->dim->value;
+			}
+		);
+
+		self::assertSameSize($properties, $hydratePropertyNames);
+		self::assertSameSize($properties, $extractPropertyNames);
+		self::assertCount(0, array_diff($properties, $hydratePropertyNames));
+		self::assertCount(0, array_diff($properties, $extractPropertyNames));
+	}
+
+	private function findConstructor(Class_ $class) : array {
+		return array_filter(
+			$class->stmts,
+			static function (Node $node): bool {
+				return $node instanceof ClassMethod && '__construct' === $node->name->name;
+			}
+		);
+	}
+
+	private function findAssignedPropertyNames(Node $node, string $callbackName, string $variableName, callable $mapper)
+	{
+		$finder = new NodeFinder();
+		$callbacks = $finder->find($node, function(Node $node) use ($callbackName) {
+			return $node instanceof Assign
+				&& $node->var->var->name instanceof Node\Identifier
+				&& $node->var->var->name->name === $callbackName;
+		});
+
+		$found = [];
+		foreach($callbacks as $callback) {
+			/** @noinspection SlowArrayOperationsInLoopInspection */
+			$found = array_merge($found, $finder->find($callback, function(Node $node) use ($variableName) {
+				return $node instanceof Assign
+					&& is_string($node->var->var->name)
+					&& $node->var->var->name === $variableName;
+			}));
+		}
+
+		return array_map($mapper, $found);
+	}
+
+	/**
+	 * @return Node[]
+	 */
+	public function propertyAstProvider(): array
+	{
+		$parser = (new ParserFactory())->create(ParserFactory::ONLY_PHP7);
+
+		$className = UniqueIdentifierGenerator::getIdentifier('Foo');
+		$classCode = 'class '.$className.' { private $bar; private $baz; protected $tab; '
+			.'protected $tar; }';
+
+		eval($classCode);
```

Comment:
Hmm, interesting, I thought I made this library kinda side-effect free :|

Guess I'll have more work to do.